### PR TITLE
Dynamic secret backup & restore

### PR DESCRIPTION
**Pull request by rooftopcellist**
#### Summary

Issue: https://github.com/ansible/awx-operator/issues/322

This approach takes in a list of secrets and iterates over them to dump the secret name and data into a structured yaml file that can be included as variables at restore time.  The secrets and their data are nested under a single secret name (`secrets`) to avoid overwriting any variables unintentionally (as is currently an issue with the `kind` variable).  

For example:
```
secrets:
  secret_key_secret:
    data: {secret_key: anJhU09NeGIyWjR1MHdxeWdWaXFRVGJwZ2FYNTZ5ZlE=}
    name: awx-secret-key
  admin_password_secret:
    data: {password: VWRsbUEzTzBQZ09XdGphUGJUZU15VlRIZ1I0Z25jRDc=}
    name: awx-admin-password
  broadcast_websocket_secret:
    data: {secret: NnMzV25mUms1TElKRGo3QkJHMHJMZjdIVUNYTVpaeUU=}
    name: awx-broadcast-websocket
  postgres_configuration_secret:
    data: {database: YXd4, host: YXd4LXBvc3RncmVz, password: OGlEcERLU2lVY3JIWm43M3FmQ2ZScmJla2tBczZOeEc=,
        port: NTQzMg==, type: bWFuYWdlZA==, username: YXd4}
    name: awx-postgres-configuration
  route_tls_secret:
    data: {ca.crt: b3B0aW9uYWwtY2EtY2VydA==, tls.crt: Y2VydC12YWx1ZQ==, tls.key: a2V5LXZhbHVl}
    name: awx-route-tls
```

The following secrets are now backed up:
* route_tls_secret
* ldap_cacert_secret
* image_pull_secret

#### Additional Info

As an extension of this, we should consider adding the awx_spec variables in the same yaml file to simplify things.  This would also allow us to remove the `_kind` fix [here](https://github.com/ansible/awx-operator/blob/devel/roles/restore/tasks/deploy_awx.yml#L3-L5).

### CODE REVIEW

**Comment by rooftopcellist** [comment]
@Spredzy I tested restoring to the same namespace to see if there are any conflicts with the secrets.  Other than the default openshift secrets, no new secrets are created, the same set of secrets will get used for both the old and new deployments of AWX.  

Fresh deployment `awx`:

```
[chadams@chadams-work awx-operator]$ oc get secrets
NAME                            TYPE                                  DATA   AGE
awx-app-credentials             Opaque                                3      16h
awx-broadcast-websocket         Opaque                                1      16h
awx-dockercfg-h47qj             kubernetes.io/dockercfg               1      16h
awx-ldap-cacert                 Opaque                                1      16h
awx-route-tls                   Opaque                                3      16h
awx-secret-key                  Opaque                                1      16h
awx-token-72kf4                 kubernetes.io/service-account-token   4      16h
awx-token-vwrcs                 kubernetes.io/service-account-token   4      16h
builder-dockercfg-r67g9         kubernetes.io/dockercfg               1      16h
builder-token-2bdxc             kubernetes.io/service-account-token   4      16h
builder-token-gr2tc             kubernetes.io/service-account-token   4      16h
chadams-test-dockercfg-gw7xj    kubernetes.io/dockercfg               1      16h
chadams-test-token-9c8pz        kubernetes.io/service-account-token   4      16h
chadams-test-token-dgx8n        kubernetes.io/service-account-token   4      16h
custom-postgres-configuration   Opaque                                6      16h
default-dockercfg-f8tf6         kubernetes.io/dockercfg               1      16h
default-token-nk8qh             kubernetes.io/service-account-token   4      16h
default-token-sbhpb             kubernetes.io/service-account-token   4      16h
deployer-dockercfg-vfbkl        kubernetes.io/dockercfg               1      16h
deployer-token-njpj6            kubernetes.io/service-account-token   4      16h
deployer-token-nndb9            kubernetes.io/service-account-token   4      16h
my-admin-password               Opaque                                1      16h
my-secret-key                   Opaque                                1      16h
[chadams@chadams-work awx-operator]$ oc get secrets | wc -l
24
```

After backup & restore to the same namespace:

```
[chadams@chadams-work awx-operator]$ oc get secrets
NAME                            TYPE                                  DATA   AGE
awx-app-credentials             Opaque                                3      16h
awx-broadcast-websocket         Opaque                                1      16h
awx-dockercfg-h47qj             kubernetes.io/dockercfg               1      16h
awx-ldap-cacert                 Opaque                                1      16h
awx-route-tls                   Opaque                                3      16h
awx-secret-key                  Opaque                                1      16h
awx-token-72kf4                 kubernetes.io/service-account-token   4      16h
awx-token-vwrcs                 kubernetes.io/service-account-token   4      16h
builder-dockercfg-r67g9         kubernetes.io/dockercfg               1      16h
builder-token-2bdxc             kubernetes.io/service-account-token   4      16h
builder-token-gr2tc             kubernetes.io/service-account-token   4      16h
chadams-test-dockercfg-gw7xj    kubernetes.io/dockercfg               1      16h
chadams-test-token-9c8pz        kubernetes.io/service-account-token   4      16h
chadams-test-token-dgx8n        kubernetes.io/service-account-token   4      16h
custom-postgres-configuration   Opaque                                6      16h
default-dockercfg-f8tf6         kubernetes.io/dockercfg               1      16h
default-token-nk8qh             kubernetes.io/service-account-token   4      16h
default-token-sbhpb             kubernetes.io/service-account-token   4      16h
deployer-dockercfg-vfbkl        kubernetes.io/dockercfg               1      16h
deployer-token-njpj6            kubernetes.io/service-account-token   4      16h
deployer-token-nndb9            kubernetes.io/service-account-token   4      16h
my-admin-password               Opaque                                1      16h
my-secret-key                   Opaque                                1      16h
test-restore-app-credentials    Opaque                                3      7m8s
test-restore-dockercfg-slrsm    kubernetes.io/dockercfg               1      7m7s
test-restore-token-76hbk        kubernetes.io/service-account-token   4      7m7s
test-restore-token-lctf8        kubernetes.io/service-account-token   4      7m7s

[chadams@chadams-work awx-operator]$ oc get secrets | wc -l
28
```

File: roles/backup/tasks/secrets.yml [hunk 1]
```diff
@@ -1,65 +1,33 @@
...
-  set_fact:
-    database_password: "{{ _postgres_configuration['resources'][0]['data']['password'] | b64decode }}"
-    database_username: "{{ _postgres_configuration['resources'][0]['data']['username'] | b64decode }}"
-    database_name: "{{ _postgres_configuration['resources'][0]['data']['database'] | b64decode }}"
-    database_port: "{{ _postgres_configuration['resources'][0]['data']['port'] | b64decode }}"
-    database_host: "{{ _postgres_configuration['resources'][0]['data']['host'] | b64decode }}"
-
-- name: Template secrets into yaml
+- name: Create Temporary secrets file
+  tempfile:
```
Comment (Zokormazo) [inline roles/backup/tasks/secrets.yml:4]: tempfile just calls mktemp. Can we garbage collect them later? i don't like leaving stale files around there

Comment (rooftopcellist) [reply roles/backup/tasks/secrets.yml:4]: _checks `/tmp` directory_, good point!  I'll add a task to garbage collect those, thanks @Zokormazo

Comment (rooftopcellist) [reply roles/backup/tasks/secrets.yml:4]: @Zokormazo I just fixed this and squashed all commits.

**Review by Zokormazo — APPROVED** [review]

**Pull request merged**